### PR TITLE
Cohortless Pilot

### PR DESCRIPTION
**Pull request by ltello**
### Context

This PR merges together 4 pieces of work cohortless-related:
- Modified journey for SITs to be able to roll-over their training programme to next academic year
- Redefinition of statuses of participants 
- Permanent cohort setup for ECTs based on induction start date from their DQT record.
- School dashboard redesign to be fully cohortless for SITs.

See individual pull requests:

  - [roll-over LP/DP partnership for next academic year](https://github.com/DFE-Digital/early-careers-framework/pull/3336)
  - [permanent cohort setup for ECTs](https://github.com/DFE-Digital/early-careers-framework/pull/3265)
  - [cohortless dashboard](https://github.com/DFE-Digital/early-careers-framework/pull/3307)

### CODE REVIEW

File: app/serializers/appropriate_bodies/induction_records_serializer.rb
Comment (peteryates) [inline app/serializers/appropriate_bodies/induction_records_serializer.rb:1]: This will probably need a quick look by a @DFE-Digital/lead-provider-cpd - the status derived differently.

File: app/serializers/appropriate_bodies/induction_records_serializer.rb
Comment (peteryates) [inline app/serializers/appropriate_bodies/induction_records_serializer.rb:48]: Status derived differently here too cc @DFE-Digital/lead-provider-cpd

File: app/serializers/delivery_partners/participants_serializer.rb
Comment (peteryates) [inline app/serializers/delivery_partners/participants_serializer.rb:17]: And here cc @DFE-Digital/lead-provider-cpd

File: app/serializers/appropriate_bodies/induction_records_serializer.rb
Comment (cwrw) [reply app/serializers/appropriate_bodies/induction_records_serializer.rb:1]: For this and below, we've chatted to Matt around those changes and those serialisers are not used in the API but the AB/DP UI for csvs, so should be ok. We also discussed the status, and there shouldn't be any changes either

File: app/serializers/appropriate_bodies/induction_records_serializer.rb
Comment (cwrw) [reply app/serializers/appropriate_bodies/induction_records_serializer.rb:48]: see other comment

File: app/serializers/delivery_partners/participants_serializer.rb
Comment (cwrw) [reply app/serializers/delivery_partners/participants_serializer.rb:17]: see other comment

**Review by peteryates — COMMENTED** [review]
I've been through this in a decent amount of detail but not line-by-line. I'm happy with the overall direction and happy there's not really much overlap with the API work. I'll take another look once it's green and let's get it merged.

File: app/controllers/schools/cohort/setup_controller.rb
Comment (tonyheadford) [inline app/controllers/schools/cohort/setup_controller.rb:4]: Hmm not sure this is a great module name. I'm guessing that this caused issues in the code (hence the addition of `::` to `Cohort` when referring to the model?

File: app/forms/schools/add_participants/base_wizard.rb
Comment (ltello) [reply app/forms/schools/add_participants/base_wizard.rb]: Reverted after rename aboved

File: app/helpers/application_helper.rb
Comment (leoapost) [inline app/helpers/application_helper.rb:42]: I see this method is only used in controllers, so I perhaps it should be a service class and add some specs?

File: app/helpers/schools/dashboard_helper.rb
Comment (leoapost) [inline app/helpers/schools/dashboard_helper.rb:10]:
```suggestion
      school_cohorts.sum { |sc| sc.current_induction_records.count }
```

The induction records have one participant, so I think counting the `current_induction_records` should be enough.

Also, Not sure if displaying the ects count is hugely important, so I wonder if it worth to add some specs to ensure the method is counting the `current_induction_records`. Someone could easily be confused and change the method to sum the `induction_records` or `active_induction_records`, which would count more induction records than intended.

File: app/services/form_data/data_store.rb
Comment (leoapost) [inline app/services/form_data/data_store.rb:44]: I can see any references to this method. Is it needed?

File: app/helpers/application_helper.rb
Comment (ltello) [reply app/helpers/application_helper.rb]: Yes, you are right. I just placed it there to speed up bug fixes. 
I have now extracted it on to its own service.
Thanks @leoapost

File: app/helpers/schools/dashboard_helper.rb
Comment (ltello) [reply app/helpers/schools/dashboard_helper.rb:10]: This scope (.ects) is to pick only current induction records belonging to an ECT.

File: app/services/form_data/data_store.rb
Comment (ltello) [reply app/services/form_data/data_store.rb:44]: Not sure where it is used.
Maybe Tony needed it for debugging purposes?

Comment (ltello) [reply app/services/form_data/data_store.rb:44]: Or what is me?
Such a long time that I hardly remember.
Let's remove it. We'll find out if it is needed somewhere.

File: spec/features/schools/training_dashboard/manage_no_ect_training_spec.rb
Comment (leoapost) [inline spec/features/schools/training_dashboard/manage_no_ect_training_spec.rb:6]: Is this a typo?

File: app/controllers/schools/participants_controller.rb
Comment (ltello) [reply app/controllers/schools/participants_controller.rb]: Haha, You are reviewing quicker than the test runner!!

**Review by leoapost — APPROVED** [review]

File: app/services/dashboard/latest_manageable_cohort.rb
Comment (tonyheadford) [inline app/services/dashboard/latest_manageable_cohort.rb:15]: I might be missing something but I don't know if this could not be summed up as:
```ruby
pilot?  ? Cohort.active_registration_cohort : Cohort.current
```
Won't that always be the case?  The `Cohort.latest` worries me a bit for what it will do in test scenarios or if things are setup ahead of time in the future, it could get misused and it's only needed for this method?

File: app/views/schools/cohort_setup/_training_confirmation_school_funded_fip.html.erb
Comment (tonyheadford) [inline app/views/schools/cohort_setup/_training_confirmation_school_funded_fip.html.erb:7]:
```suggestion
                                no_visited_state: true %>, approved by the DfE.
```

File: app/services/dashboard/latest_manageable_cohort.rb
Comment (ltello) [reply app/services/dashboard/latest_manageable_cohort.rb:15]: Bear in mind that this is computing a min. 
The `Cohort.latest` thing is just useful in cases (specs) where the current cohort has not been created yet, therefore, returning the most recent cohort created rather than returning nil and breaking the test.
In prod, we will always have the Cohort.current created. If not, it would return the previous.

Comment (tonyheadford) [reply app/services/dashboard/latest_manageable_cohort.rb:15]: `Cohort.current` should only change when a new cohort is added with a `registration_start_date` and an `academic_year_start_date` that takes over from the previous one?

Comment (ltello) [reply app/services/dashboard/latest_manageable_cohort.rb:15]: `Cohort.current` is only based on the current date.

Comment (tonyheadford) [reply app/services/dashboard/latest_manageable_cohort.rb:15]: Thanks for clarifying @ltello, I still don't think this should be here especially if it's just a "hack" to fix specs.  My main concern is `Cohort.latest` as I don't think that should exist because I think it could too easily be misunderstood and we have the necessary methods we need in `current`,`previous` and `next` on `Cohort`.  I don't want to hold up this PR though but can you add a tech debt ticket to get rid of this and correct the specs that might be missing a Cohort creation and/or `travel_to` blocks to work correctly with the cohorts used in those specs?

**Review by tonyheadford — APPROVED** [review]
Great job @ltello! Thanks for sticking with it

**Pull request merged**
